File: blender_extension/roblox_interface/roblox_server/format_types/array.py
import struct

format = struct.Struct("<L")
format_size = format.size
unpack_from = format.unpack_from
pack = format.pack

def read(self, format_data, **masks):
	list_len, = unpack_from(*self.read_buffer(format_size))
	list_value = []
	for _ in range(list_len):
		v = self.parse(format_data["value"], **masks)
		list_value.append(v if len(v) > 1 else v[0])
	return (list_value,)

def write(self, args, args_count, format_data, **masks):
	list_value = args[args_count]
	list_len = len(list_value)
	self.write_buffer(pack(list_len), format_size)
	if list_len > 0:
		values_type = type(list_value[0])
		if (values_type is list) or (values_type is tuple):
			for v in list_value:
				self.parse(v, 0, format_data["value"], **masks)
		else:
			for v in list_value:
				self.parse((v,), 0, format_data["value"], **masks)
	return args_count + 1

File: roblox_plugin/src/Server/formatTypes/array.ts
//!native
import * as ProcessFormats from "../processFormats";
import SendMessagesThread from "../sendMessages";
import ReceiveMessagesThread from "../receiveMessages";

interface ArrayFormatData {
	type: "array";
	value: ProcessFormats.FormatData;
}
const DATA_SIZE = 4;
export function read(
	receiveThread: ReceiveMessagesThread,
	args: defined[],
	formatData: ArrayFormatData,
	masks: Map<string, boolean>,
) {
	const [readBuffer, readOffset] = ReceiveMessagesThread.readBuffer(receiveThread, DATA_SIZE);
	const arraySize = buffer.readu32(readBuffer, readOffset);
	const array = new Array(arraySize);
	for (const i of $range(0, arraySize - 1)) {
		const v: defined[] = [];
		ReceiveMessagesThread.parse(receiveThread, v, formatData.value, masks);
		array[i] = v.size() > 1 ? v : v[0];
	}
	args.push(array);
}
export function write(
	sendThread: SendMessagesThread,
	args: defined[],
	argsCount: number,
	formatData: ArrayFormatData,
	masks: Map<string, boolean>,
): number {
	const array = args[argsCount] as defined[];
	const arraySize = array.size();
	const data = buffer.create(DATA_SIZE);
	buffer.writeu32(data, 0, arraySize);
	SendMessagesThread.writeBuffer(sendThread, data, DATA_SIZE);
	if (arraySize > 0) {
		if (typeIs(array[0], "table")) {
			for (const v of array) {
				SendMessagesThread.parse(sendThread, v as defined[], 0, formatData.value, masks);
			}
		} else {
			for (const v of array) {
				SendMessagesThread.parse(sendThread, [v], 0, formatData.value, masks);
			}
		}
	}
	return argsCount + 1;
}
